Configurable Event Logging

ejoerns opened the pull request:
This is a first draft implementation of the event logging feature noted in concept issue #1070

jluebbe commented on src/event-log.c:
```diff
@@ -0,0 +1,402 @@
+	/* Close open stream */
+	g_clear_pointer(&logger->logstream, g_object_unref);
+
+	/* determine number of existing rotation files */
+	for (guint num = 0; num < G_MAXUINT; num++) {
+		g_autofree gchar *tryfile = g_strdup_printf("%s.%d", logger->filename, num);
+
+		if (g_file_test(tryfile, G_FILE_TEST_EXISTS)) {
+			existing_files++;
+		} else {
+			break;
+		}
+	}
```
Can this be dropped by always trying to rotate `maxfiles` files and ignoring `ENOENT` on `rename`?

ejoerns replied:
I've thought about this, too. But have not found a proper solution for not having maxfiles set. Or there must be a default maxfiles of like 100 files or so.

jluebbe replied:
I'd say a default of 10 (so 1…9) would be fine. That would keep the size fluctuation low enough.

ejoerns replied:
10 would be fine for me. Currently maxfiles is defined as the number of rotation files to keep (thus would be 1..10). Do you think this should be changed?

jluebbe replied:
I would have included the currently active file in the count. So maxfiles would not be exceeded in total.

ejoerns replied:
As an alternative: In `logrotate` the option is called `rotate` and a `rotate` of `0` means "old versions are removed rather than rotated". But `-1` there means "old logs are not removed at all" (the ability we just removed) so we probably shouldn't adapt to its naming anyway.

For `maxfiles` it does not feel that clear how it can be interpreted. But maybe counting this as the total number of files (as you suggested) is fine.

ejoerns replied:
Changed the code now to interpret `max-files` as the maximum of all files (including the logfile itself).

jluebbe commented on src/event-log.c:
```diff
@@ -0,0 +1,402 @@
+	g_clear_pointer(&logger->name, g_free);
+	g_clear_pointer(&logger->filename, g_free);
+	g_clear_pointer(&logger->logstream, g_object_unref);
+
+
```
Redundant empty line.

jluebbe commented on src/event-log.c:
```diff
@@ -0,0 +1,402 @@
+		logfile = g_file_new_for_path(logger->filename);
+		logger->logstream = g_file_append_to(logfile, G_FILE_CREATE_NONE, NULL, &ierror);
+		if (!logger->logstream) {
+			g_warning("Failed opening log file for appending: %s", ierror->message);
```
"Failed to open..."

jluebbe commented on src/event-log.c:
```diff
@@ -0,0 +1,402 @@
+	rotate_logfiles(logger, strlen(output));
+
+	if (!g_output_stream_write_all(G_OUTPUT_STREAM(logger->logstream), output, strlen(output), &written, NULL, &ierror)) {
+		g_warning("Failed writing log file: %s", ierror->message);
```
"Failed to write..."

jluebbe commented on qemu-test-rauc-config:
```diff
@@ -5,6 +5,10 @@ grubenv=/tmp/boot/grub/grubenv
 data-directory=/tmp/rauc
 
+[log.testlogger]
+filename=event.log
```
Should this be somewhere in `/tmp`? If not, where does this end up?

ejoerns replied:
default for relative paths is currently relative to data-directory. Thus `/tmp/rauc/event.log` here.

jluebbe replied:
Perhaps a comment here would be nice.

ejoerns commented on src/event-log.c:
```diff
@@ -0,0 +1,405 @@
+ * Simply converts all log fields into a json object.
+ *
+ * | {
+ * |   "date": "2023-06-14T21:15:41Z"
+ * |   "MESSAGE" : "Booted into rootfs.0 (A)",
+ * |   "MESSAGE_ID" : "e60e0add-d345-4cb8-b796-eae0d497af96",
```
Different in which way? It's just an example with two entries. Could be shortened to one.

ejoerns replied:
Ok, misunderstood you. So the reason for having "date" as it is is just that i wasn't sure if its the right name and I first of all wanted to differentiate it from the other entries since it is the only one generated in the log function itself.

Thus could be changed to `"DATE"` or `"TIMESTAMP"`.

jluebbe replied:
Is it useful to save some space in the JSON output? If so, perhaps `"TS"` and `"MSG(_ID)"`?

ejoerns replied:
I used the `journalctl -o json` as a reference which also just plainly prints out the internal names of structured logging.

ejoerns commented:
Changes made:

* log files are now counted `log.1` to `log.<max-files>` instead of starting with `log.0`
* `max-files` now defaults to `10`
* non-existing log files will just be skipped during rotation (simplified code)
* `logger` was renamed to `loggers`
* added comment to location of testlogger in qemu test config

jluebbe commented on src/config_file.c:
```diff
@@ -157,6 +158,141 @@ gboolean parse_bundle_formats(guint *mask, const gchar *config, GError **error)
+						error,
+						G_KEY_FILE_ERROR,
+						G_KEY_FILE_ERROR_INVALID_VALUE,
+						"Unsupported event type '%s'", logger->events[j]);
```
```suggestion
						"Unsupported event log type '%s'", logger->events[j]);
```

jluebbe replied:
This still applies to the current version.

jluebbe commented on src/config_file.c:
```diff
@@ -157,6 +158,141 @@ gboolean parse_bundle_formats(guint *mask, const gchar *config, GError **error)
+			return FALSE;
+		}
+
+		logger->maxfiles = key_file_consume_integer(key_file, *group, "max-files", &ierror);
```
We should probably check that the provided value is in a sensible range (i.e. 1..1000).

jluebbe replied:
This still applies to the current version.

ejoerns replied:
We've already ensured "> 1" and I've now added a check for "< 1000"

jluebbe commented on src/context.c:
```diff
@@ -325,6 +326,15 @@ static gboolean r_context_configure_target(GError **error)
 	}
 
+	/* set up logging */
+	if (context->config->loggers) {
+		for (GList *l = context->config->loggers; l != NULL; l = l->next) {
```
The if should not be needed, as the loop condition checks for NULL as well.

jluebbe commented on docs/reference.rst:
```diff
@@ -697,6 +697,51 @@ No built-in slot update will run and no hook will be executed.
+  * ``install`` - Logs start and end of installation
+  * ``boot`` - Logs boot information
+  * ``mark`` - Logs slot marking information
+  * ...
```
This should also mention `all`.

jluebbe commented on src/event-log.c:
```diff
@@ -0,0 +1,398 @@
+
+static const gchar *supported_event_types[] = {
+	"all",
+	R_EVENT_LOG_TYPE_BOOT,
+	R_EVENT_LOG_TYPE_INSTALL,
+	R_EVENT_LOG_TYPE_SERVICE,
+	R_EVENT_LOG_TYPE_WRITE_SLOT,
+	R_EVENT_LOG_TYPE_BOOT_SELECTION,
```
Perhaps some comments to clarify the respective expected content?

jluebbe replied:
These are `#define`d in `include/event-log.h`, so the comments should go there. Still applies to the current version.

ejoerns replied:
Without actually using it, the scope is probably still a bit vague but tried to be a bit clearer.

Might need some adaptions and extensions when adding actual log message printouts later anyway.

jluebbe commented on src/event-log.c:
```diff
@@ -0,0 +1,398 @@
+	now_formatted = g_date_time_format(now, "%Y-%m-%dT%H:%M:%SZ");
+
+	json_builder_begin_object(builder);
+	json_builder_set_member_name(builder, "date");
```
```suggestion
	json_builder_set_member_name(builder, "TS");
```

jluebbe replied:
This now uses `TIMESTAMP`, but not `TS`. My intention there was to *reduce* the verbosity and size of the log keys a bit.

ejoerns replied:
I tried to stay close to what systemd journal json output would provide you, but since this uses far longer names anyway (like `_SOURCE_MONOTONIC_TIMESTAMP`, it might not be relevant here, anyway).

jluebbe commented on src/event-log.c:
```diff
@@ -0,0 +1,398 @@
+	g_autoptr(JsonBuilder) builder = json_builder_new();
+
+	now = g_date_time_new_now_utc();
+	now_formatted = g_date_time_format(now, "%Y-%m-%dT%H:%M:%SZ");
```
Perhaps use `g_date_time_format_iso8601()`, as this also provides sub-second precision? That could be useful to correlate events if as lot is happening.

ejoerns replied:
I chose the format as this is consistent with the existing ones.

Another thing is that `g_date_time_format_iso8601()` requires glib 2.62 as a minimal version.

If we agree that it is worth the effort, we can either bump glib minimal version or conditionally import the method.

2.62 is available since Yocto/poky `dunfell`, Ubuntu 20.04, Debian oldstable (`bullseye`)

jluebbe replied:
As the current format is also ISO8601 (just without microsecond precision), we can use this for now and move to `g_date_time_format_iso8601()` later.

jluebbe commented on src/event-log.c:
```diff
@@ -0,0 +1,398 @@
+		g_autofree gchar *to_file = g_strdup_printf("%s.%d", logger->filename, file_num);
+
+		/* skip non-existing files */
+		if (!g_file_test(from_file, G_FILE_TEST_EXISTS))
+			continue;
+
+		if (g_rename(from_file, to_file) == -1) {
+			int err = errno;
+			g_warning("Failed to rotate log file: %s", g_strerror(err));
+		}
```
Avoid the initial check and just handle `errno==ENOENT`.

jluebbe replied:
This still applies to the current version.

jluebbe commented on src/event-log.c:
```diff
@@ -0,0 +1,398 @@
+		return;
+
+	/* Close open stream */
+	g_clear_pointer(&logger->logstream, g_object_unref);
```
Should we flush before closing? Or even `fsync()`?

ejoerns replied:
From the Glib `GOutputStream` documentation:

> Closing the stream will implicitly cause a flush.

Since a `GFileOutputStream` is also a `GOutputStream`, this will apply here, too.

However, calling `fsync()` should be possible when assuming having a `GUnixOutputStream` as we do for the update handlers already.

jluebbe commented on src/event-log.c:
```diff
@@ -0,0 +1,398 @@
+	rotfile = g_strdup_printf("%s.1", logger->filename);
+	if (g_rename(logger->filename, rotfile) == -1) {
+		int err = errno;
+		g_warning("Failed to rotate log file: %s", g_strerror(err));
+		return;
```
In this case, we no longer have a log file open, right? So writing the next log event would fail?

ejoerns replied:
Two possible solutions I see here:

a) ignore that we failed to rename and continue with re-opening (the probably still existing) log file

b) abort and disable logger

the disadvantage of a) is that it will probably fail on each further iteration and exceeds `max-size`. The disadvantage of b) is that we do not log anymore despite we probably could.

ejoerns replied:
Done so.

jluebbe commented on src/event-log.c:
```diff
@@ -0,0 +1,398 @@
+	rotate_logfiles(logger, strlen(output));
+
+	if (!g_output_stream_write_all(G_OUTPUT_STREAM(logger->logstream), output, strlen(output), &written, NULL, &ierror)) {
+		g_warning("Failed to write log file: %s", ierror->message);
```
Hmm, this will keep generating `g_warnings` for any further message. Perhaps we should have a `broken` flag per logger and drop any messages for broken loggers?

ejoerns replied:
Added `broken` flag now (+test).

jluebbe commented on src/event-log.c:
```diff
@@ -0,0 +1,398 @@
+		REventLogger* logger = l->data;
+
+		/* Filter out by event type */
+		if ((g_strcmp0(logger->events[0], "all") != 0) &&
```
Do we ensure that `all` is always the first type in this array?

ejoerns replied:
We now ensure that `"all"` cannot be combined with other values (and will be set as default if no value is provided).

jluebbe commented on src/event-log.c:
```diff
@@ -0,0 +1,398 @@
+		g_autoptr(GFileInfo) logfile_info = NULL;
+
+		logfile = g_file_new_for_path(logger->filename);
+		logfile_info = g_file_query_info(logfile, G_FILE_ATTRIBUTE_STANDARD_SIZE, G_FILE_QUERY_INFO_NONE, NULL, &ierror);
```
Couldn't this be done using `g_seekable_tell()` after opening the file?

ejoerns replied:
Already opening the logfile in the setup probably makes sense anyway. Since `GFileOutputStream` implements `GSeekable` this should work, yes.

jluebbe reviewed (changes requested):
Still missing is a test case for the `system.conf` parsing.

jluebbe commented on docs/reference.rst:
```diff
@@ -697,6 +697,51 @@ No built-in slot update will run and no hook will be executed.
+
+``max-files``
+  Configures the maximum number of rotation files to keep.
+  E.g. if set to ``3``, only ``<filename>.0``, ``<filename>.1`` and
+  ``<filename>.2`` will be kept during rotation.
+  Defaults to 10 if unset.
```
This is no longer correct.

ejoerns replied:
The mentioning of `.0` is invalid, yes. The default of 10 however was just added and should be valid.

jluebbe commented on docs/reference.rst:
```diff
@@ -697,6 +697,51 @@ No built-in slot update will run and no hook will be executed.
+
+**[log.<logger>] sections**
+
+With a logger section, a *RAUC event logger* can be configured.
```
We should point to the section in `advanced.rst` here for the general behavior.

ejoerns replied:
Indeed, this did not exist yet when I started here, will fix.

jluebbe commented on docs/advanced.rst:
```diff
@@ -1594,3 +1594,56 @@ https://github.com/caruhome/upparat
+  output-format=json
+  max-size=1M
+  max-files=5
+
```
We should specify what happens if rotation or logging fails.

ejoerns replied:
Yes. I was not quite sure about this, yet. But maybe marking the logger broken as suggested above could be a way to handle this.

jluebbe replied:
```suggestion
If an error occurs during logging (such as disk full or write errors), that logger is marked as broken and no longer used.
An ongoing installation is **not** aborted.
```

ejoerns commented on docs/advanced.rst:
```diff
@@ -1594,3 +1594,56 @@ https://github.com/caruhome/upparat
+
+.. note:: All events logged using the internal event logging framework will
+   also be available on the journal.
+   Some of them also come with journal catalog entries to provide more details.
```
Yes, but I am not sure if this note should make it into this PR since adding the catalog is a follow-up for now.

ejoerns commented:
Changes made:

* Add some internal module documentation for event-log
* Fix the above-noted documentation issues
* mention ``all`` as valid event type in doc
* replaced `"bootsel"` event type by `"mark"`
* added a couple of config parsing tests
* made `filename` option mandatory and fixed error handling
* let `logger->events` default to `{"all", NULL}` if no events given and error if `"all"` is combined with others in config
* fix freeing of `logger->events`
* added GError handling to `rotate_logfiles()`
* first of all do not abort on failing `g_rename()`
* added `logger->broken` `gboolean` flag and support deactivating logger on log file opening failure

ejoerns commented:
Changes made:

* fix freeing of `config->loggers`
* put chunk introducing `config->loggers` into appropriate commit

ejoerns commented on include/event-log.h:
```diff
@@ -0,0 +1,88 @@
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(REventLogger, r_event_log_free_logger);
+
+/**
+ * Simple Logging convenience method.
```
More or less for reference. I am still unsure if it should be used at all.

ejoerns replied:
It's not used by the added structured logging tests.

jluebbe replied:
So, should it be removed or remain untested?

ejoerns replied:
My fault. The above command should have been "It's **now** used by the added structured logging tests."

ejoerns commented:
Changes made:

* added cross-link from `reference.rst` to `advanced.rst` chapter
* removed note of *catalog* from docs first of all (and made description more precise)
* Changed behavior of `max-files` to include the logfile itself in the count. Adapted tests.
* Added error handling, sanity checks and tests for ensuring `max-files > 0`
* changed `"date"` to `"TIMESTAMP"` in json output
* fixed broken logger handling
* open logger already during `r_event_log_setup_logger()` and use `g_seekable_tell()` for determining log file size

jluebbe commented on src/event-log.c:
```diff
@@ -0,0 +1,406 @@
+		if (!g_file_test(from_file, G_FILE_TEST_EXISTS))
+			continue;
+
+		if (g_rename(from_file, to_file) == -1) {
```
Add a comment that this will replace the .<maxfiles-1> file.

jluebbe commented on test/event-log.c:
```diff
@@ -0,0 +1,299 @@
+
+	logger = g_new0(REventLogger, 1);
+	logger->name = g_strdup("testlogger");
+	logger->filename = g_strdup("/this/path/does/not/exist/testfile.log");
```
Another case would be a full device. You can use `/dev/full` for that:
```
# echo foo > /dev/full 
-bash: echo: write error: No space left on device
```

ejoerns replied:
Added test case "/event-log/log-writer/no-space-left" for this.

jluebbe commented on docs/reference.rst:
```diff
@@ -697,6 +697,54 @@ No built-in slot update will run and no hook will be executed.
+
+``filename``
+  The log file name used for logging.
+  If not absolute path is given, the location is assumed to be relative to the
```
```suggestion
  If no absolute path is given, the location is assumed to be relative to the
```

jluebbe commented on docs/reference.rst:
```diff
@@ -697,6 +697,54 @@ No built-in slot update will run and no hook will be executed.
+  configuration error.
+
+``events``
+  Colon-separated list of events to log. Currently supported event types are:
```
```suggestion
  Semicolon-separated list of events to log. Currently supported event types are:
```

jluebbe commented on docs/reference.rst:
```diff
@@ -697,6 +697,54 @@ No built-in slot update will run and no hook will be executed.
+``max-size``
+  Allows to configure a basic log rotation.
+  When given, the logger's log file will be rotated before reaching
+  the size configured with ``max-size`` and renamed to ``<filename>.0``.
```
```suggestion
  the size configured with ``max-size`` and renamed to ``<filename>.1``.
```

jluebbe commented on docs/reference.rst:
```diff
@@ -697,6 +697,54 @@ No built-in slot update will run and no hook will be executed.
+  Allows to configure a basic log rotation.
+  When given, the logger's log file will be rotated before reaching
+  the size configured with ``max-size`` and renamed to ``<filename>.0``.
+  Existing rotation file names will be incremented by one.
```
```suggestion
  Existing rotation file names will be incremented by one.
  The oldest file is removed.
```

jluebbe commented on src/config_file.c:
```diff
@@ -157,6 +158,174 @@ gboolean parse_bundle_formats(guint *mask, const gchar *config, GError **error)
+					error,
+					G_KEY_FILE_ERROR,
+					G_KEY_FILE_ERROR_INVALID_VALUE,
+					"Value for 'max-files' muste be >= 1");
```
```suggestion
					"Value for 'max-files' must be >= 1");
```

jluebbe reviewed (changes requested):
Some comments from the previous review are still open.

ejoerns commented:
Changes made:

* Added and fixed documentation pieces as requested
* Fixed typos
* Added all so-far-known fields to formatter
* Made code comments clearer where requested
* Added test for setting up logger on full device
* Added tests for actually testing the code through structured logging (includes actually testing filtering)

ejoerns commented:
Changes made:

* Added two missing doc suggestions
* Simply error-out on failed log rotation (as requested)

jluebbe commented on src/config_file.c:
```diff
@@ -157,6 +158,141 @@ gboolean parse_bundle_formats(guint *mask, const gchar *config, GError **error)
+	for (gchar **group = groups; *group != NULL; group++) {
+		GError *ierror = NULL;
+		g_autoptr(REventLogger) logger = NULL;
+		gchar *logger_name;
```
Still applies to current version.

ejoerns commented:
* Added comments to log type `#define`s
* Added comments to `REventLogFormat` enum
* Made `logger_name` const pointer
* Added check for `max-files` being <= 1000 in config parsing
* Remove not needed `if` guard when iterating over loggers in `context.c`
* Renamed `"TIMESTAMP"` to shorter `"TS"` in json log output
* Combined file existence check with `g_rename` handling by ignoring `err == ENOENT`

ejoerns commented:
Rebased onto current `master`.

jluebbe commented on src/event-log.c:
```diff
@@ -0,0 +1,405 @@
+	if (g_rename(logger->filename, rotfile) == -1) {
+		int err = errno;
+		g_set_error(error, G_FILE_ERROR, g_file_error_from_errno(err), "Failed to rotate log file: %s", g_strerror(err));
+		return FALSE;
+	} else {
+		logger->filesize = 0;
+	}
```
It doesn't really make sense to have an `else` here after the error return above.

The `filesize` reinit belongs in the re-open part below.

ejoerns replied:
Yes, absolutely makes sense.

jluebbe commented on src/event-log.c:
```diff
@@ -9,6 +9,20 @@
+	R_EVENT_LOG_TYPE_MARK,
+	R_EVENT_LOG_TYPE_INSTALL,
+	R_EVENT_LOG_TYPE_SERVICE,
+	R_EVENT_LOG_TYPE_WRITE_SLOT,
```
This needs to be `NULL` terminated for use with `g_strv_contains`.

jluebbe reviewed (changes requested):
We currently don't use dashes (but underscores) in filenames. Perhaps rename `event-log.[ch]`? Looks good otherwise.

ejoerns commented on include/event-log.h:
```diff
@@ -0,0 +1,123 @@
+ * * "BUNDLE_HASH" : The bundle hash associated with the log message
+ * * "BUNDLE_VERSION" : The bundle version associated with the log message
+ * * "TRANSACTION_ID" : The transaction ID associated with the log message
+ * * "RAUC_SLOT" : The slot name associated with the log message
```
A bit historically because at first, I was unsure how the fields should be named.
Probably `SLOT_NAME` fits better here.

ejoerns commented:
* Renamed  `event-log.[ch]` to `event_log.[ch]`
* Renamed `RAUC_SLOT` to `SLOT_NAME`
* Moved resetting of `logger->filesize` to log file re-opening
* Added `NULL` sentinel to `supported_event_types`

jluebbe commented:
The cross build failure is unrelated. It seems the multiarch containers are outdated.
